File: 36- Longest common subsequence/Python/solution5.py
# Space-optimized dynamic programming solution:

# Time complexity: O(nm)
# Space complexity: O(m)

def lcs(str1, str2):
  n = len(str1)
  m = len(str2)
  dp = [0] * (m+1)
  tempDp = [0] * (m+1)
  for i in range(1, n+1):
    for j in range(1, m+1):
      if str1[i-1] == str2[j-1]:
        tempDp[j] = 1 + dp[j-1]
      else:
        tempDp[j] = max(tempDp[j-1], dp[j])
    for j in range(1, m+1):
      dp[j] = tempDp[j]
      tempDp[j] = 0
  return dp[m]

File: 36- Longest common subsequence/C++/solution5.cpp
// Space-optimized dynamic programming solution:

// Time complexity: O(nm)
// Space complexity: O(m)

int lcs(string str1, string str2){
  int n = str1.length();
  int m = str2.length();
  int dp[m+1];
  int tempDp[m+1];
  for(int i = 0; i < m+1; i++) dp[i] = tempDp[i] = 0;
  for(int i = 1; i < n+1; i++){
    for(int j = 1; j < m+1; j++){
      if(str1[i-1] == str2[j-1])
        tempDp[j] = 1 + dp[j-1];
      else
        tempDp[j] = max(tempDp[j-1], dp[j]);
    }
    for(int j = 1; j < m+1; j++){
      dp[j] = tempDp[j];
      tempDp[j] = 0;
    }
  }
  return dp[m];
}


